### ISSUE TITLE
[Regression] Role/claim changes in identity aren't persisted into the auth cookie

### ISSUE THREAD

**Issue by kanadaj** [aspnet/Security#1788]
When ASP.NET Core Identity tries to update the current user's claims, it updates the Principal of the current request, however the update to the claims is not persisted to the cookie.

I've also made an issue for this in aspnet/Identity before I've managed to figure out the details: https://github.com/aspnet/Identity/issues/1845

I'm not sure which repo this issue really belongs to considering this seems like an interop issue.

Repro repo with steps: https://github.com/kanadaj/aspnetrefreshrolesrepro/tree/master

**Comment by kanadaj** [comment]
@Tratcher here you go. I still think the problem is that Identity modifies `context.Principal` but Security persists `context.Ticket.Principal`; a quick verification of this would be welcome since I'm running a system in production and it's causing issues for the users.

**Comment by kanadaj** [comment]
https://github.com/aspnet/Security/blob/1df139eb6d8ef1c1e74a14dd20c9dd13eb6e0e23/test/Microsoft.AspNetCore.Authentication.Test/CookieTests.cs#L958
This test doesn't cover the case where ctx.ReplacePrincipal() is used to replace the principal with a new one, since you're using the reference directly.

**Comment by kanadaj** [comment]
As a quick and dirty workaround, this seems to work
```
options.OnRefreshingPrincipal = context =>
{
    // Any original code in OnRefreshingPrincipal
    // ...

    // Copy the new principal claims into the old value to keep the reference to context.ticket.Principal
    var oldIdentity = context.CurrentPrincipal.Identities.First();
    var newIdentity = context.NewPrincipal.Identities.First();

    Claim claim;
    while ((claim = oldIdentity.Claims.FirstOrDefault()) != null)
    {
        oldIdentity.RemoveClaim(claim);
    }

    foreach (var newClaim in newIdentity.Claims)
    {
        oldIdentity.AddClaim(newClaim);
    }

    return Task.FromResult(0);
};
```

**Comment by Tratcher** [comment]
@HaoK can you take a look? If this regressed then it's a patch candidate.

**Comment by HaoK** [comment]
Sure, I wouldn't be surprised if this was a regression from the cloning change

**Comment by HaoK** [comment]
Well so this is interesting, looks like the security stamp might have been relying on the old behavior before to persist updates to the cookie:

See https://github.com/aspnet/Identity/blob/dev/src/Identity/SecurityStampValidator.cs#L80

Expectation was that ReplacePrincipal + ShouldRenew would cause the cookie to be updated for the user.  I'll add a functional test in identity confirming that this is indeed broken now

**Comment by HaoK** [comment]
This is the regression change with the cloning, might need to just revert?  https://github.com/aspnet/Security/commit/1df139eb6d8ef1c1e74a14dd20c9dd13eb6e0e23

Since it does change the semantics of ReplacePrincipal (which Identity's SecurityStampValidator was relying on)

**Comment by kanadaj** [comment]
@HaoK Just wondering here, why not keep a reference to the ticket instead and set up `context.Principal` for the `CookieValidatePrincipalContext` to point to the ticket's Principal? Or would that break other things?

**Comment by HaoK** [comment]
So certainly we can work around the new behavior in identity, but if this broke identity, it should at least be looked at like a regression.

**Comment by kanadaj** [comment]
Or alternatively, alter ReplacePrincipal to also set the ticket's principal.

**Comment by kanadaj** [comment]
That's true

**Comment by Tratcher** [comment]
@kanadaj may be right, we should be able to fix ReplacePrincipal to allow explicit cookie updates and still clone to prevent later app contamination.

**Comment by Tratcher** [comment]
@muratg this needs triaging. It's a regression in 2.1 and may warrant a patch.

**Comment by muratg** [comment]
cc @Eilon

**Comment by Eilon** [comment]
@HaoK - are we good to close this bug now? Is the fix in the patch and in 2.2 and in master?

**Comment by HaoK** [comment]
Yep, thx for reminding me, fixed via https://github.com/aspnet/Security/pull/1811
